refactor: fix mypy errors in data_sources/*

We're in the middle of trying to improve the type hints we use in cve-bin-tool.  I previously filed some issues as found with pyright, this particular set of issues is done using mypy.  We're new to type checking, so it's entirely possible that the types currently in the file are incorrect or incomplete. To test the file yourself, you'll need to install mypy (e.g. `pip install mypy`) and then run it with the filename you want to test.   You may need to also run `mypy --install-types` to get it to work on some files.

Note that I'm filing issues for each file individually to help folk who are looking for fun stuff to do as part of hacktoberfest so you can claim one and not interfere with anyone else working on another file.  If you've already got your 4 commits for hacktoberfest, please leave these issues for someone who's still working on theirs!

Here's a list of issues found:

```
cve_bin_tool/data_sources/curl_source.py:131: error: Item "None" of "Optional[Match[str]]" has no attribute "group"
cve_bin_tool/data_sources/osv_source.py:42: error: Need type annotation for "all_cve_entries" (hint: "all_cve_entries: List[<type>] = ...")
cve_bin_tool/data_sources/gad_source.py:46: error: Need type annotation for "all_cve_entries" (hint: "all_cve_entries: List[<type>] = ...")
cve_bin_tool/data_sources/rsd_source.py:44: error: Need type annotation for "all_cve_entries" (hint: "all_cve_entries: List[<type>] = ...")
```

Helpful resources:
* [Python typing documentation](https://docs.python.org/3/library/typing.html)
* [mypy cheat sheet](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)
* If you can't find the answer in the docs, googling the error messages above will often help!
* [cve-bin-tool's contributor docs](https://github.com/intel/cve-bin-tool/blob/main/CONTRIBUTING.md) 
* If you've contributed to open source but not this project, you might just want our [checklist for a great pull request](https://github.com/intel/cve-bin-tool/blob/main/CONTRIBUTING.md#checklist-for-a-great-pull-request)

Tiny tips for new contributors:
* cve-bin-tool uses https://www.conventionalcommits.org/ style for commit messages, and we have a test that checks the title of your PR.  A good potential title for this one is in the title of this issue.
* You can make an issue auto close by including a comment "fixes #ISSUENUMBER" in your PR comments where ISSUENUMBER is the actual number of the issue.

--- a/cve_bin_tool/data_sources/curl_source.py
+++ b/cve_bin_tool/data_sources/curl_source.py
@@ -127,10 +127,13 @@ class Curl_Source(Data_Source):
         Return the versions we have Curl data for.
         """
         regex = re.compile(r"curlcve-(\d+.\d+.\d).json")
-        return [
-            regex.search(filename).group(1)
-            for filename in glob.glob(str(Path(self.cachedir) / "curlcve-*.json"))
-        ]
+        versions = []
+        for filename in glob.glob(str(Path(self.cachedir) / "curlcve-*.json")):
+            match = regex.search(filename)
+            if match:
+                version = match.group(1)
+                versions.append(version)
+        return versions
 
     def get_cve_list(self):
         self.affected_data = []

--- a/cve_bin_tool/data_sources/gad_source.py
+++ b/cve_bin_tool/data_sources/gad_source.py
@@ -6,6 +6,7 @@ import datetime
 import io
 import zipfile
 from pathlib import Path
+from typing import List
 
 import aiohttp
 import yaml
@@ -43,7 +44,7 @@ class GAD_Source(Data_Source):
         self.gad_commit_url = self.GAD_COMMIT_URL
         self.gad_compare_url = self.GAD_COMPARE_URL
         self.gad_api_url = self.GAD_API_URL
-        self.all_cve_entries = []
+        self.all_cve_entries: List[dict] = []
 
         self.session = None
 

--- a/cve_bin_tool/data_sources/osv_source.py
+++ b/cve_bin_tool/data_sources/osv_source.py
@@ -7,6 +7,7 @@ import io
 import json
 import zipfile
 from pathlib import Path
+from typing import List
 
 import aiohttp
 from cvss import CVSS3
@@ -39,7 +40,7 @@ class OSV_Source(Data_Source):
 
         self.osv_url = self.OSV_URL
         self.gs_url = self.OSV_GS_URL
-        self.all_cve_entries = []
+        self.all_cve_entries: List[dict] = []
 
         self.session = None
 

--- a/cve_bin_tool/data_sources/rsd_source.py
+++ b/cve_bin_tool/data_sources/rsd_source.py
@@ -7,6 +7,7 @@ import io
 import json
 import zipfile
 from pathlib import Path
+from typing import List
 
 import aiohttp
 from cvss import CVSS2, CVSS3
@@ -41,7 +42,7 @@ class RSD_Source(Data_Source):
         self.rsd_commit_url = self.RSD_COMMIT_URL
         self.rsd_compare_url = self.RSD_COMPARE_URL
         self.rsd_api_url = self.RSD_API_URL
-        self.all_cve_entries = []
+        self.all_cve_entries: List[dict] = []
 
         self.session = None
 

